CRM/Contribute/BAO/Query.php

```diff
@@ -258,6 +258,7 @@ public static function whereClauseSingle(&$values, &$query) {
       case (strpos($name, '_amount') !== FALSE):
       case (strpos($name, '_date') !== FALSE && $name != 'contribution_fulfilled_date'):
       case 'contribution_campaign_id':
+      case 'contribution_cancel_reason':
```

Comment:
Question: Why do we need to add case for every contribution field of type varchar? Doesn't the default section responsible to handle this automatically?

Reply:
the **default:** section doesn't take care properly of string fields. If you see the result query, it adds where clause `civicrm_contribution.financial_type_id IN ("3", "1",  "4","2")` when it's not necessary, and then the **IN** operator it's not parsed properly

Basically, I added cancel_reason in the same *case* block as other string field are parsed properly, like **contribution_source**

Reply:
I agree the default section should work / be fixed. The switch statement mostly exists because of bad coding historically. 

-We should really get a test on it too.

Reply:
ok, I agree as well that the proper fix is to correct **default:** code block. But for that fix  I'm afraid I won't be able to take the lead to make a new PR
Refactoring that, might big impact on the whole Contribution Query class, and ever piece of code / feature that's related with it, and I don't think I'm aware of how to fix it properly